systemdspawner/systemdspawner.py

```diff
@@ -147,25 +149,37 @@ class SystemdSpawner(Spawner):
         """
     ).tag(config=True)
 
+    # Characters that are safe for systemd units.
+    safe_chars = set(string.ascii_lowercase + string.digits + string.ascii_uppercase + '_-')
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # All traitlets configurables are configured by now
         self.unit_name = self._expand_user_vars(self.unit_name_template)
 
         self.log.debug('user:%s Initialized spawner with unit %s', self.user.name, self.unit_name)
 
-    def _expand_user_vars(self, string):
+    def _escape_variable(self, in_string):
+        """
+        Escape variables for systemd unit naming
+        """
+        return escapism.escape(in_string, safe=self.safe_chars, escape_char='-')
+
+    def _expand_user_vars(self, in_string):
         """
         Expand user related variables in a given string
 
         Currently expands:
           {USERNAME} -> Name of the user
           {USERID} -> UserID
+          {SERVERNAME} -> Name of the server (self.name)
         """
-        return string.format(
-            USERNAME=self.user.name,
-            USERID=self.user.id
-        )
+        # Strip the trailing - if we don't have a name.
+        return in_string.format(
+            USERNAME=self._escape_variable(self.user.name),
+            USERID=self.user.id,
+            SERVERNAME=self._escape_variable(self.name)
+        ).rstrip('-')
```

Comment:
I don't think this fully solves the problem, since the user configured template might already contain unsafe characters. It might be easier to just escape the whole thing instead of the username and servername separately.